devise を使ってユーザー認証を実装する

yumiya-g opened the pull request:
サンプルアプリにユーザー認証機能を実装しました。

yumiya-g committed:
Add: Gemを追加しました

- devise-i18n: deviseを多言語化
- letter_opener_web: メールテスト

yumiya-g committed:
Add: deviseインストールを実行して、マニュアルセットアップを行いました

yumiya-g committed:
Add: ルーティング調整とログイン制限処理を実装しました

- routes.rb: ルートページ（'/'）アクセスすると、Books一覧が表示される
- application_controller.rb: 未ログイン時にアクセスすると、ログインページへリダイレクトする

yumiya-g committed:
Add: ヘッダーメニューとログアウトボタンを実装しました

yumiya-g committed:
Add: usersコントローラーを生成しました（index,showのみ）

yumiya-g committed:
Add: Deviseにカスタマイズしたusersコントローラーとビューファイルを生成しました

yumiya-g committed:
Add: usersコントローラーのルーティングを設定しました

yumiya-g committed:
Move: ヘッダーテンプレートを共通フォルダに移動しました

yumiya-g committed:
Add: 未ログイン時と、ログインエラー発生時にアラートメッセージを表示させました

yumiya-g committed:
Fix: ヘッダーメニューを共通化しました

- ユーザーの一覧とログアウトのアンカーテキストは、訳文を参照している

yumiya-g committed:
Add: ユーザーの一覧ページを実装しました

- 登録ID順で表示する

yumiya-g committed:
Fix: ユーザー編集ページを実装しました

- Userモデルに、住所と自己紹介文のカラムを追加
- ユーザーの一覧ページとユーザー詳細ページのテンプレート修正

yumiya-g committed:
Add: ページネーションを追加しました

- kaminariを使って、1ページあたりのユーザー表示数を「2」に設定

yumiya-g committed:
Add: Gem「letter_opener_web」を導入しました

- パスワード再設定時、Deviseからの送信メールをブラウザ上で確認するため

yumiya-g committed:
Add: Bookモデルのビューテンプレートにヘッダーを追加しました

- layouts/headerパーシャル読み込み
- 詳細ページの枠の上部に余白を設定

yumiya-g committed:
Fix: Deviseが生成するテンプレートファイルの翻訳対応をしました

yumiya-g committed:
Fix: ログインページの翻訳対応を行いました

yumiya-g committed:
Fix: erblintを実行して、エラー箇所を修正しました

yumiya-g committed:
Fix: rubocopを実行して、エラーが箇所を修正しました

yumiya-g committed:
Delete: 使用していないファイルを削除しました

yumiya-g committed:
Fix: 言語を日本語に設定して、翻訳ファイルを修正しました

yuuu commented on app/controllers/users/confirmations_controller.rb:
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class Users::ConfirmationsController < Devise::ConfirmationsController
```
今回の仕様だと `confirmation` や `unlock` 関連のファイルは不要ですので削除しておいてください 🙏

yumiya-g replied:
ご指摘いただきありがとうございます。不要ファイルの削除を行いました。

yuuu commented on app/controllers/users/passwords_controller.rb:
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+class Users::PasswordsController < Devise::PasswordsController
```
このファイルも特に手を加えないのであれば削除して問題ないはずです 👍

yumiya-g replied:
こちらのファイルも削除を実行いたしました。

yuuu commented on app/models/user.rb:
```diff
@@ -0,0 +1,10 @@
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+
+  paginates_per 2
```
念のため確認ですが `2` で問題なかったですか？

yumiya-g replied:
ページ分割を確認するため「2」としておりましたが、seed.rbで初期ユーザーを20件登録するよう変更しましたので、こちらは「10」に変更しました❗

yuuu commented on app/views/books/index.html.erb:
```diff
@@ -1,3 +1,4 @@
+<%= render '/layouts/header' %>
```
こちらは `views/layouts/application.html.erb` に移動させると何か問題がありますか？

yumiya-g replied:
` views/layouts/application.html.erb`に記載すべきでした。
ヘッダーメニューは`application.html.erb`で表示させ、他のテンプレートファイルからの読み込みを削除しました。

yuuu commented on app/views/devise/mailer/email_changed.html.erb:
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @email %>!</p>
+
+<% if @resource.try(:unconfirmed_email?) %>
+  <p>We're contacting you to notify you that your email is being changed to <%= @resource.unconfirmed_email %>.</p>
```
devise周りのファイルについて全体的にi18n対応が漏れているようなので対応をお願いいたします。

yumiya-g replied:
ご指摘いただきありがとうございます。不要なファイルを削除後、メーラー本文のi18n対応を行いました。

yuuu commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,313 @@
+  # Configure the e-mail address which will be shown in Devise::Mailer,
+  # note that it will be overwritten if you use your own mailer class
+  # with default "from" parameter.
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
ここはdeviseによって送信されるメールの `from` に当たる部分なので `no-reply@example.com` のようなメールアドレスに変えておくと良いですね。

yumiya-g replied:
ご指摘いただきありがとうございます。
こちらは`no-reply@example.com`に変更をしました。

yuuu commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,313 @@
+
+  # ==> Configuration for :validatable
+  # Range for password length.
+  config.password_length = 6..128
```
昨今のセキュリティ情勢を踏まえると、最小文字数6文字は少ない気がするので8文字以上としてはどうでしょうか？

yumiya-g replied:
こちらもご指摘の通り、最低文字数が「6」では少ないと感じましたので、8文字に変更しました。

yumiya-g committed:
Add: ユーザー一覧ページでの表示件数を10件に変更しました

- seeds.rb:migrate実行時に、ユーザー20件を追加する処理を追加

yumiya-g committed:
Fix: 送信元メールアドレスの修正と、パスワードの最小文字数を「6」->「8」に変更しました

yumiya-g committed:
Fix: ヘッダーメニュー用テンプレートの読み込み処理を修正しました

- 「layouts/application.html.erb」でパーシャル『header』を読み込む

yumiya-g committed:
Delete: 未使用のファイルを削除しました

- UserモデルのControllerから削除
- Deviseから生成されたファイルを削除

yumiya-g committed:
Delete: 未使用のファイルを削除しました

- UserモデルのControllerから削除
- Deviseから生成されたファイルを削除
- `devise_for`で設定されていた`users/sessions`コントローラーを削除

yumiya-g committed:
Fix: パスワード再発行時に送信されるメール本文のi18n対応を行いました

yuuu commented on app/controllers/users/sessions_controller.rb:
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+class Users::SessionsController < Devise::SessionsController
+  # before_action :configure_sign_in_params, only: [:create]
```
何も手を加えないのであればこのファイルも削除してしまって大丈夫ですね 👍

yumiya-g replied:
`routes.rb`で、このファイルを読み込む設定になっていました。
読み込み設定を解除してファイルを削除しました。

yuuu commented on app/views/devise/registrations/edit.html.erb:
```diff
@@ -0,0 +1,73 @@
+  </div>
+
+  <% if devise_mapping.confirmable? && resource.pending_reconfirmation? %>
+    <div>Currently waiting confirmation for: <%= resource.unconfirmed_email %></div>
```
細かいですがここのi18n対応が漏れていますね 👀

yumiya-g replied:
`devise-i18n`でViewファイルを生成して対応しました。

yuuu commented on app/views/layouts/_header.html.erb:
```diff
@@ -0,0 +1,10 @@
+<header>
+  <h1><%= link_to "Books App", books_path %></h1>
```
ここもi18n対応が漏れていますね 👀

yumiya-g replied:
`translation_ja.yml`にタイトルを設定しました。

yuuu commented on app/views/layouts/application.html.erb:
```diff
@@ -11,6 +11,11 @@
   <body class="<%= params[:action] %>">
+    <% if user_signed_in? %>
+      <%= render 'layouts/header' %>
+    <% else %>
```
この書き方だと、ログイン中はalertが表示されなくなってしまいますが問題ないでしょうか？

yumiya-g replied:
問題がありそうでしたので、`notice`と`alert`の出力処理を調整しました。

yuuu commented on app/views/users/index.html.erb:
```diff
@@ -0,0 +1,17 @@
+<p style="color: green"><%= notice %></p>
```
こちらも `alert` と同様に `application.html.ern` に記述してはどうでしょうか？

yumiya-g replied:
アドバイスいただきましたように、`application.html.erb`にまとめました。

yumiya-g committed:
Fix: `alert`と`notice`の表示を共通化しました

- `alert`と`notice`は、ログイン・未ログインに関わらず表示するよう変更
- `alert`と`notice`は、`true`の場合に文字列を出力する
- 各テンプレートに記述した`notice`を削除

yumiya-g committed:
Fix: Gem`devise-i18n`でViewファイルを再生成しました

- `devise/registrations/edit.html.erb`：見出しh2タグの出力は、初期値ではなく`User.model_name.human`に変更
- `app/views/layouts/_header.html.erb`：アプリのタイトルをi18n対応

yumiya-g committed:
Fix: `erblint`,`rubocop`を実行してエラーを解消しました

yumiya-g committed:
Merge branch 'my-user_page' of github.com:yumiya-g/fjord-books_app-2023 into my-user_page
